pkg/controller/rotation/handle_secrets_rotate.go

```diff
@@ -0,0 +1,417 @@
+func (c *Controller) importExistingSecretFromEnv(ctx context.Context, typ database.SecretType, secretName, envvar, envval string, mutators ...importMutatorFunc) ([]*database.Secret, error) {
+	logger := logging.FromContext(ctx).Named("importExistingSecretFromEnv").
+		With("type", typ).
+		With("env", envvar).
+		With("secretName", secretName)
+
+	now := time.Now().UTC()
+
+	// allSecretsBytes is the ordered list of all secrets in bytes format.
+	allSecretsBytes := make([][]byte, 0, 4)
+
+	// Iterate over each secret in the envvar, separated by commas. This collects
+	// the raw secret values as bytes.
+	parts := strings.Split(envval, ",")
+	for _, part := range parts {
+		// These should be in the form secret://...?target=file. Transform them
+		// back into their normal form.
+		ref := strings.TrimSpace(part)
+		if !strings.HasPrefix(ref, "secret://") {
+			continue
+		}
+		ref = strings.TrimPrefix(ref, "secret://")
+		ref = strings.TrimSuffix(ref, "?target=file")
+
+		logger.Infow("importing secret value", "ref", ref)
+
+		// Get the initial secret value.
+		val, err := c.secretManager.GetSecretValue(ctx, ref)
+		if err != nil {
+			return nil, fmt.Errorf("failed to access initial secret %s: %w", ref, err)
+		}
+
+		// v1 secrets are base64-encoded separated by commas.
+		insideParts := strings.Split(val, ",")
+		for i, insidePart := range insideParts {
+			dec, err := base64util.DecodeString(insidePart)
+			if err != nil {
+				return nil, fmt.Errorf("failed to decode initial secret %s at %d: %w", ref, i, err)
+			}
+			allSecretsBytes = append(allSecretsBytes, dec)
+		}
+	}
+
+	// Apply any mutations.
+	for _, fn := range mutators {
+		var err error
+		allSecretsBytes, err = fn(allSecretsBytes)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	// Now that we have all bytes, iterate and create the secrets.
+	orderedRefs := make([]string, 0, len(allSecretsBytes))
+	for _, b := range allSecretsBytes {
+		parent := path.Join(c.config.SecretsParent, secretName)
+		version, err := c.secretManager.CreateSecretVersion(ctx, parent, b)
+		if err != nil {
+			return nil, fmt.Errorf("failed to add secret version to %s: %w", parent, err)
+		}
+		orderedRefs = append(orderedRefs, version)
+	}
+
+	// Now that all secrets are created upstream, create the database records.
+	created := make([]*database.Secret, 0, len(orderedRefs))
+	for i, ref := range orderedRefs {
+		logger.Infow("create secret reference in database", "ref", ref)
+
+		secret := &database.Secret{
+			Type:      typ,
+			Reference: ref,
+			Active:    true,
+
+			CreatedAt: now.Add(time.Duration(-i) * time.Hour),
+			UpdatedAt: now.Add(time.Duration(-i) * time.Hour),
+		}
+		if err := c.db.SaveSecret(secret, RotationActor); err != nil {
+			return nil, fmt.Errorf("failed to save secret ref %s in database: %w", ref, err)
+		}
+		created = append(created, secret)
+	}
+
+	return created, nil
+}
+
+// mutateCookieKeysSecrets handles the special case of importing cookie keys.
```

Comment:
Background: In order to keep the cookie HMAC key and cookie encryption key pairs in sync, I'm merging them into a new, single secret.